### PR TITLE
Add support for custom version requirements with a predicate

### PR DESCRIPTION
**Pull request by jspahrsummers**
This enables support for requirements like "pin to this named branch," where maybe a semantic version isn't available.

### CODE REVIEW

**Comment by jspahrsummers** [comment]
@kastiglione @mdiep Can you please take a look at this and see if it makes any sense? The logic ended up a bit different from Carthage, because Arbiter doesn't inherently have knowledge of what kinds of _versions_ should clobber others—instead, it can only check a version against each requirement.

File: include/arbiter/Requirement.h [hunk 1]
```diff
@@ -32,9 +39,39 @@ typedef enum
...
+   * the requirement.
+   *
+   * This result is unique in that it _overrides other requirements_, which is
+   * sometimes desirable (e.g., pinning to a specific named branch or tag
+   * instead of a semantic version).
+   *
+   * If this result is returned multiple times for different versions, it is
+   * unspecified which version will be selected.
+   */
+  ArbiterRequirementSuitabilityBestPossibleChoice,
```
Comment (mdiep) [inline include/arbiter/Requirement.h]: _Ideal_ might be a better name here.

File: include/arbiter/Requirement.h [hunk 2]
```diff
@@ -32,9 +39,39 @@ typedef enum
...
+  /**
+   * The version is suitable for (satisfies) the requirement.
+   */
+  ArbiterRequirementSuitabilitySuitable,
+  
+  /**
+   * The version should be considered the best possible choice for satisfying
+   * the requirement.
+   *
+   * This result is unique in that it _overrides other requirements_, which is
```
Comment (kastiglione) [inline include/arbiter/Requirement.h]: Not to get caught up on naming, but I like the word "override", or something that communicates that this has been hand picked.

**Comment by kastiglione** [comment]
Makes sense, though that doesn't say much since my knowledge about the inner workings of Carthage, or dependency managers in general, pales compared to you two.

**Comment by mdiep** [comment]
The other thing to consider here is that Carthage has a bug currently where it loops endless in some cases because of these sort of overriding constraints with non-semantic-versions. Because they're compatible with everything, you can get into situations where a version is picked that doesn't meet your requirements, even though nothing explicitly selected that version.

**Comment by jspahrsummers** [comment]
> Because they're compatible with everything, you can get into situations where a version is picked that doesn't meet your requirements, even though nothing explicitly selected that version.

I think this implementation would behave differently, because versions determined to be the "best" are not just compatible with everything—they _override_ everything, and other versions get pruned from the graph.

That said, I'm not very happy with how this turned out, and given that neither of you were very gung-ho either, this probably deserves more thought. 😔

**Comment by mdiep** [comment]
I think Arbiter should be opinionated about how to handle non-semantic versions.

**Comment by jspahrsummers** [comment]
What should that mean here? It's hard to be opinionated when the library is this general. :confused:

**Comment by mdiep** [comment]
I just suspect that there's 1 reasonable approach to versioning with unversioned dependencies (non-semantic tags, branches, commits, etc.). IMO Arbiter should implement that so there's 1 easy built-in way, even if it enables other approaches.

**Pull request closed without merging**